src/lib/support/tests/TestBytesCircularBuffer.cpp

```diff
@@ -0,0 +1,191 @@
+    NL_TEST_ASSERT(inSuite, buffer.IsEmpty());
+    NL_TEST_ASSERT(inSuite, buffer.Push(ByteSpan(s, 7)) == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, !buffer.IsEmpty());
+}
+
+void TestPushWrap(nlTestSuite * inSuite, void * inContext)
+{
+    uint8_t storage[10];
+    const uint8_t s1[] = { '1', '2', '3', '4', '5', '6', '7', '8', '9', '0' };
+    const uint8_t s2[] = { 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j' };
+    uint8_t o[sizeof(s1)];
+    MutableByteSpan readSpan(o);
+
+    // Total space = 9
+    BytesCircularBuffer buffer(storage, sizeof(storage));
+    NL_TEST_ASSERT(inSuite, buffer.IsEmpty());
+
+    // Used = 6, Free = 3
+    NL_TEST_ASSERT(inSuite, buffer.Push(ByteSpan(s1, 4)) == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, !buffer.IsEmpty());
+    NL_TEST_ASSERT(inSuite, buffer.GetFrontSize() == 4);
+    NL_TEST_ASSERT(inSuite, buffer.ReadFront(readSpan) == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, memcmp(s1, readSpan.data(), 4) == 0);
+
+    // Used = 6, Free = 3, s1 was discarded due to lack of storage
+    NL_TEST_ASSERT(inSuite, buffer.Push(ByteSpan(s2, 4)) == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, !buffer.IsEmpty());
+    NL_TEST_ASSERT(inSuite, buffer.GetFrontSize() == 4);
+    NL_TEST_ASSERT(inSuite, buffer.ReadFront(readSpan) == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, memcmp(s2, readSpan.data(), 4) == 0);
+
+    // Used = 9, Free = 0
+    NL_TEST_ASSERT(inSuite, buffer.Push(ByteSpan(s1, 1)) == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, !buffer.IsEmpty());
+    NL_TEST_ASSERT(inSuite, buffer.GetFrontSize() == 4);
+    NL_TEST_ASSERT(inSuite, buffer.ReadFront(readSpan) == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, memcmp(s2, readSpan.data(), 4) == 0);
+
+    // Used = 3, Free = 6
+    NL_TEST_ASSERT(inSuite, buffer.Pop() == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, !buffer.IsEmpty());
+    NL_TEST_ASSERT(inSuite, buffer.GetFrontSize() == 1);
+    NL_TEST_ASSERT(inSuite, buffer.ReadFront(readSpan) == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, memcmp(s1, readSpan.data(), 1) == 0);
+}
+
+void TestPushWrapAtEdge(nlTestSuite * inSuite, void * inContext)
+{
+    uint8_t storage[10];
+    const uint8_t s1[] = { '1', '2', '3', '4', '5', '6', '7', '8', '9', '0' };
+    const uint8_t s2[] = { 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j' };
+    uint8_t o[sizeof(s1)];
+    MutableByteSpan readSpan(o);
+
+    // Total space = 9
+    BytesCircularBuffer buffer(storage, sizeof(storage));
+    NL_TEST_ASSERT(inSuite, buffer.IsEmpty());
+
+    // Used = 6, Free = 3
+    NL_TEST_ASSERT(inSuite, buffer.Push(ByteSpan(s1, 4)) == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, !buffer.IsEmpty());
+    NL_TEST_ASSERT(inSuite, buffer.GetFrontSize() == 4);
+    NL_TEST_ASSERT(inSuite, buffer.ReadFront(readSpan) == CHIP_NO_ERROR);
+    NL_TEST_ASSERT(inSuite, memcmp(s1, readSpan.data(), 4) == 0);
+
+    // Used = 9, Free = 0, mDataEnd wrap to 0
```

Comment:
Does it?  I'd expect `mDataEnd == 9` at this point....  Which is good, because `mDataStart == 0` here, so if `mDataEnd` were also 0 we would be violating our invariants.